wandb/sdk/launch/agent/agent.py

```diff
@@ -222,25 +228,135 @@ def loop(self) -> None:
             self.print_status()
+            self._pool.close()
+            self._pool.join()
+
+    # Threaded functions
+    def thread_run_job(
+        self,
+        launch_spec: Dict[str, Any],
+        job: Dict[str, Any],
+        default_config: Dict[str, Any],
+        api: Api,
+    ) -> None:
+
+        try:
+            self._thread_run_job(launch_spec, job, default_config, api)
+        except Exception:
+            wandb.termerror(f"{LOG_PREFIX}Error running job: {traceback.format_exc()}")
+            api.ack_run_queue_item(job["runQueueItemId"])
+
+    def _thread_run_job(
+        self,
+        launch_spec: Dict[str, Any],
+        job: Dict[str, Any],
+        default_config: Dict[str, Any],
+        api: Api,
+    ) -> None:
+        thread_id = threading.current_thread().ident
+        assert thread_id is not None
+        job_status = JobAndRunStatus()
+        with self._jobs_lock:
+            self._jobs[thread_id] = job_status
+        project = create_project_from_spec(launch_spec, api)
+        _logger.info("Fetching and validating project...")
+        project = fetch_and_validate_project(project, api)
+        _logger.info("Fetching resource...")
+        resource = launch_spec.get("resource") or "local-container"
+        backend_config: Dict[str, Any] = {
+            PROJECT_SYNCHRONOUS: False,  # agent always runs async
+        }
+
+        backend_config["runQueueItemId"] = job["runQueueItemId"]
+        _logger.info("Loading backend")
+        override_build_config = launch_spec.get("build")
+        override_registry_config = launch_spec.get("registry")
+
+        build_config, registry_config = resolve_build_and_registry_config(
+            default_config, override_build_config, override_registry_config
+        )
+        builder = load_builder(build_config)
+
+        default_runner = default_config.get("runner", {}).get("type")
+        if default_runner == resource:
+            backend_config["runner"] = default_config.get("runner")
+        backend = load_backend(resource, api, backend_config)
+        backend.verify()
+        _logger.info("Backend loaded...")
+        run = backend.run(project, builder, registry_config)
+
+        if not run:
+            with self._jobs_lock:
+                job_status.failed_to_start = True
+            return
+        with self._jobs_lock:
+            job_status.run = run
+        while self._jobs_event.is_set():
+            if self._check_run_finished(job_status):
+                return
+            time.sleep(AGENT_POLLING_INTERVAL)
+        # temp: for local, kill all jobs. we don't yet have good handling for different
+        # types of runners in general
+        if isinstance(run, LocalSubmittedRun):
+            run.command_proc.kill()
+
+    def _check_run_finished(self, job_status: JobAndRunStatus) -> bool:
+        if job_status.run is None:
+            if job_status.failed_to_start:
+                return True
+            return False
```

Comment:
this is an interesting bit of logic lmao. If not run, but job didn't start, run finished...

Reply:
I'll add a comment